### PR TITLE
Problem: no support for contracts

### PR DESCRIPTION
**Pull request by splix**
Basic support for contracts (list, add)

### CODE REVIEW

**Comment by splix** [comment]
fix #13

File: src/contracts.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
+#![cfg_attr(feature = "dev", feature(plugin))]
+#![cfg_attr(feature = "dev", plugin(clippy))]
+
+#![deny(clippy, clippy_pedantic)]
+#![allow(missing_docs_in_private_items, unknown_lints)]
+
```
Comment (dulanov) [inline src/contracts.rs]: Probably unneccesary, usually are declared on top level (lib.rs or main.rs).

File: src/contracts.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+extern crate serde;
+extern crate serde_json;
+extern crate glob;
+extern crate futures;
+
+use self::glob::glob;
+use self::serde_json::Value;
+pub use keystore::Address;
+use std::fs::File;
```
Comment (dulanov) [inline src/contracts.rs]: `use self::glob::glob` > just `use glob::glob`
`pub use keystore::Address` > `use keystore::Address`

File: .travis.yml [hunk 3]
```diff
@@ -9,7 +9,7 @@ os:
   - osx
 before_script:
   - |
-    (travis_wait cargo install rustfmt || true)
+    (travis_wait cargo install rustfmt --force || true)
 script:
```
Comment (dulanov) [inline .travis.yml]: Ah and oh... always travis...

File: src/contracts.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        match File::open(path) {
+            Ok(f) => serde_json::from_reader(f).or(Err(ContractError::IO)),
+            Err(_) => Err(ContractError::IO),
+        }
+    }
+
+    /// List all available contracts
+    pub fn list(&self) -> Vec<Value> {
+        let files = glob(&format!("{}/*.json", &self.dir.to_str().unwrap())).unwrap();
+        files.filter(|x| x.is_ok())
```
Comment (dulanov) [inline src/contracts.rs]: Need to check for Windows, better to use PathBuf to constract platfrom specific paths.

File: src/contracts.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            .map(|x| Contracts::read_json(x.unwrap().as_path()))
+            .filter(|x| x.is_ok())
+            .map(|x| x.unwrap())
```
Comment (dulanov) [inline src/contracts.rs]: There are a lot of `unwrap()`, they all are exactly execute without errors?

File: src/contracts.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+            .collect()
+    }
+
+    /// Validate contract structure
+    pub fn is_valid(&self, contract: &Value) -> Result<(), ContractError> {
```
Comment (dulanov) [inline src/contracts.rs]: `is_` > `Result<_>`?

File: Cargo.toml [hunk 7]
```diff
@@ -34,6 +34,8 @@ rustc-serialize = "0.3"
 serde = "0.9"
 serde_json = "0.9"
 serde_derive = "0.9"
+glob = "0.2.11"
+tempdir = "0.3.5"
```
Comment (dulanov) [inline Cargo.toml]: You you sure to fix minor version?

File: src/contracts.rs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+extern crate futures;
+
+use self::glob::glob;
+use self::serde_json::Value;
+pub use keystore::Address;
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
+/// Contracts Service
+pub struct Contracts {
```
Comment (dulanov) [inline src/contracts.rs]: `#[derive(Debug, Clone, Copy, PartialEq, Eq)]`?

File: src/contracts.rs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            None => return Err(ContractError::InvalidContract),
+        };
+        if !addr.is_string() {
+            return Err(ContractError::InvalidContract);
+        }
+        match addr.as_str().unwrap().parse::<Address>() {
+            Ok(_) => {}
+            Err(_) => return Err(ContractError::InvalidContract),
+        }
+        return Ok(());
```
Comment (dulanov) [inline src/contracts.rs]: `Ok(())` is more idiomatic

File: src/contracts.rs [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            .filter(|x| x.is_ok())
+            .map(|x| x.unwrap())
+            .collect()
+    }
+
+    /// Validate contract structure
+    pub fn is_valid(&self, contract: &Value) -> Result<(), ContractError> {
+        if !contract.is_object() {
+            return Err(ContractError::InvalidContract);
+        }
```
Comment (dulanov) [inline src/contracts.rs]: I prefer to insert empty lines between multiline statements.

File: src/contracts.rs [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        return Ok(());
+    }
+
+    /// Add new contract to storage
+    pub fn add(&self, contract: &Value) -> Result<(), ContractError> {
+        self.is_valid(contract)?;
+        let addr = contract.get("address")
+            .unwrap()
+            .as_str()
+            .unwrap();
```
Comment (dulanov) [inline src/contracts.rs]: hm, critic, better to unwrap explicitly with `expect("Expect ...")`

File: src/contracts.rs [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        let filename = format!("{}/{}.json", &self.dir.to_str().unwrap(), addr);
```
Comment (dulanov) [inline src/contracts.rs]: Duplicate logic, was before.

File: src/lib.rs [hunk 13]
```diff
@@ -26,15 +27,21 @@ extern crate rustc_serialize;
...
 mod request;
 mod serialize;
-
-
-use jsonrpc_core::{IoHandler, Params};
+/// Contracts stuff
+pub mod contracts;
+mod storage;
+
+use self::serde_json::Value;
```
Comment (dulanov) [inline src/lib.rs]: I think, `self` for relative referencing from child modules.

File: src/lib.rs [hunk 14]
```diff
@@ -87,8 +94,39 @@ pub fn start(addr: &SocketAddr, client_addr: &SocketAddr) {
...
+    let storage = Storages::new();
+    if !storage.init().is_ok() {
+        panic!("Unable to initialize storage")
+    }
+    let chain = ChainStorage::new(&storage, "default".to_string());
+    if !chain.init().is_ok() {
+        panic!("Unable to initialize chain")
+    }
+    let contracts_service = Arc::new(Contracts::new(chain.get_path("contracts".to_string())
+                                                        .expect("Directory for contracts")));
```
Comment (dulanov) [inline src/lib.rs]: `.expect("Expect directory for contracts")`

File: src/storage.rs [hunk 15]
```diff
@@ -0,0 +1,85 @@
+#![cfg_attr(feature = "dev", feature(plugin))]
+#![cfg_attr(feature = "dev", plugin(clippy))]
+
+#![deny(clippy, clippy_pedantic)]
+#![allow(missing_docs_in_private_items, unknown_lints)]
```
Comment (dulanov) [inline src/storage.rs]: Was before

File: src/storage.rs [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+            if log_enabled!(LogLevel::Info) {
+                info!("Init new storage at {}", self.base_dir.display());
+            }
+            fs::create_dir(self.base_dir)?
+        }
+        Ok(())
+    }
+}
+
+/// Subdir for a chain
```
Comment (dulanov) [inline src/storage.rs]: Good for comments, but not for public documentation.

File: tests/contracts_test.rs [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+#[test]
+fn should_see_all_contracts() {
+    let c = Contracts::new(contracts_path());
+    let act = c.list();
+    assert_eq!(act.len(), 1)
+}
+
+#[test]
+fn should_add_contract() {
+    let tmp_dir = TempDir::new("test-contracts").expect("create temp dir");
```
Comment (dulanov) [inline tests/contracts_test.rs]: Important to prefix all messages as "Expect...".

File: src/contracts.rs (hunk 2 above)
Comment (splix) [inline src/contracts.rs]:
```
error[E0432]: unresolved import `glob::glob`
 --> src/contracts.rs:9:5
  |
9 | use glob::glob;
  |     ^^^^^^^^^^ Did you mean `self::glob`?
```
Don't know why

File: src/contracts.rs (hunk 5 above)
Comment (splix) [inline src/contracts.rs]: should, I'm doing `x.is_ok()` before that step

File: src/contracts.rs (hunk 6 above)
Comment (splix) [inline src/contracts.rs]: hm, yeah. it was `bool` before. will figure out better name

File: src/contracts.rs (hunk 11 above)
Comment (splix) [inline src/contracts.rs]: ok. but it was just validate in previous step

File: src/contracts.rs (hunk 12 above)
Comment (splix) [inline src/contracts.rs]: not actually, previous was for listing a dir

File: Cargo.toml (hunk 7 above)
Comment (splix) [inline Cargo.toml]: ?

Comment (dulanov) [inline Cargo.toml]: you can minor version - `glob = "0.2"`

File: src/contracts.rs (hunk 5 above)
Comment (tomusdrw) [inline src/contracts.rs]: You can try https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map instead.

**Pull request merged**